Create Blog Post “how-to-send-mms-messages-with-node-js” (#2508)
Nexmo/deved-platform: content/blog/en/how-to-send-mms-messages-with-node-js.md

---
title: How to Send MMS Messages with Node.js
description: The Vonage Messages allows you to send MMS (Multimedia Messaging
  Service) from US shortcode numbers you have in your dashboard to other numbers
  within the United States. In this article you will learn how to send an MMS
  using Node.js using the official vonage-node client library. Prerequisites
  Before you start, […]
thumbnail: /content/blog/how-to-send-mms-with-node-js-dr/send-mms-nodejs.png
author: amanda-cavallaro
published: true
published_at: 2023-06-30T11:39:24.631Z
updated_at: 2023-06-30T11:39:24.655Z
category: tutorial
tags:
  - mms
  - messages-api
  - javascript
comments: true
redirect: ""
canonical: ""
---

> The Vonage Messages API now allows you to send MMS (Multimedia Messaging Service) from US shortcode numbers you have in your dashboard to other numbers within the United States. 

Reviewer: ```suggestion
The Vonage Messages API now allows you to send MMS (Multimedia Messaging Service) from US 10DLC, TFN, and Shortcode numbers you have in your dashboard to other numbers within the United States. 
```
Can now also send MMS from 10DLC and TFN numbers.

> The Vonage Messages API now allows you to send MMS (Multimedia Messaging Service) from US 10DLC, TFN, and Shortcode numbers you have in your dashboard to other numbers within the United States. 

Reviewer 2: I'll add them as further reading at the end!

In this article, you will learn how to send an MMS using Node.js using the official [vonage-node-sdk](https://github.com/vonage/vonage-node-sdk) client library.

## Prerequisites

Before you start, make sure you have the following:

* A basic understanding of JavaScript & Node.js
* Node.js installed on your machine
* A fresh US SMS & MMS capable number. You can reference our [Numbers Guide](https://developer.vonage.com/en/numbers/guides/number-management) for details on how to buy a number if you don't have one.

> **Note**: MMS messaging only works for A2P (Application to person) use cases. Therefore if you were to purchase two virtual phone numbers to send and receive the message to each other other.

Reviewer: This seems a little unclear. How about something like:
```suggestion
**Note**: MMS messaging only works for A2P (Application to Person) use cases. The Vonage Messages API cannot be used to send MMS messages from one virtual number to another virtual number.
```

<sign-up number></sign-up>

## Getting Started

Before we begin, let's get set up for success by taking care of a few things inside the Vonage dashboard.

Let's start by creating a new Messages Application. To do this, head to the [Application section](https://dashboard.nexmo.com/messages/applications) of your admin dashboard and click on *Create a new application*. 

![Your Applications page containing a purple button to create a new application](/content/blog/how-to-send-mms-messages-with-node-js/yourapplications.jpg "Your applications page")

Next, you will need to fill out the form with your application name and the required webhook URLs. 

![Messages capability showing inbound and status URLs using MockBin](/content/blog/how-to-send-mms-messages-with-node-js/messages.jpg "Message capability with added URLs ")

> Every Messages & Dispatch application requires you to specify a `Status URL` and an `Inbound URL`. In larger, production-ready applications, you would have these pointing to a URL on your own server. For this tutorial there is no need for that, you'll just need a URL that can respond with a `200 OK` status, you can use a service such as [MockBin](http://mockbin.org) to provide what you need as I did in the example above.

Reviewer: ```suggestion
Every Messages application requires you to specify a `Status URL` and an `Inbound URL`. In larger, production-ready applications, you would have these pointing to a URL on your own server. For this tutorial there is no need for that, you'll just need a URL that can respond with a `200 OK` status, you can use a service such as [MockBin](http://mockbin.org) to provide what you need as I did in the example above.
```

> Every Messages application requires you to specify a `Status URL` and an `Inbound URL`. In larger, production-ready applications, you would have these pointing to a URL on your own server. For this tutorial there is no need for that, you'll just need a URL that can respond with a `200 OK` status, you can use a service such as [MockBin](http://mockbin.org) to provide what you need as I did in the example above.

Reviewer 3: We use ngrok in most blog posts and have a lot of docs/tutorials showing how to do that. If you use mockbin I think you need to go through the steps to show how to use it, or write another mini article that shows how to use it. Something like this:

- https://developer.vonage.com/en/blog/using-ngrok-in-rails-in-2022
- https://developer.vonage.com/en/blog/ngrok-in-node-red-dr

Reviewer 2: This is a code update; the original writer used Mockbin, and I kept as is; I had never used it before, and I followed that example it was easy enough to follow to use, and personally to me, I found it cool to learn about one more new website to do this. But if you have strong opinions, I am happy to change it to use Ngrok.

Reviewer 3: I don't mind using MockBin rather than Ngrok. I just think it needs a bit more explaining here what is Mockbin and how to use it.

Also in the image, the URLs are very long and get cut off. So someone wouldn't even know that the end of the URL is different in each endpoint ( /inbound, /status).

I think if I was fairly new, I would find this section a bit confusing. But just needs like 2/3 sentences explanation.

Reviewer 2: I've added some lines explaining how to use Mockbin, @ruskibenya

![Generate Public and Private Key](/content/blog/how-to-send-mms-messages-with-node-js/keys.jpg "Generate Public and Private Key")

In order to authenticate API requests alongside your API key and API secret you will also need a public/private key pair, which can be automatically generated for you by clicking the *Generate public/private key pair* link.

> This will set a public key in the form field, and also download a `private.key` file to your machine. You will need to put this file wherever you will write the code we put together.

Reviewer 3: Change "wherever you will write the code" to "in the directory we will create shortly".

> Finish up by clicking *Create a new application.* You will then be asked to select a number to use with this application by clicking the *Link* button next to your number of choice. If you do not have any numbers on your dashboard you can purchase them there or via the command line.

Reviewer: ```suggestion
Finish up by clicking *Create a new application.* You will then be asked to select a number to use with this application by clicking the *Link* button next to your number of choice. If you do not have any numbers on your dashboard you can purchase them there or via the command line using the [Vonage CLI](https://github.com/Vonage/vonage-cli).
```

> Note down the *Application ID*, it will join your API key, API secret and Private Key in the code we will work on next.

Reviewer: @manchuck can probably weigh in here, but do you need all 4 credentials when instantiating a `Client` object using the Node SDK? If not, then we probably don't need API key and API secret since we're not using Basic Auth?

Reviewer 2: I tested before doing this, and the API key and secret were not necessary indeed. However, as it was still in the getting started example page I didn't want to remove it. Please let me know how you think I should proceed here.

Reviewer 2: Chuck Reeves
[9 minutes ago]
 yea we always want to show that to get the users into the habit of including them with using the @vonage/server-sdk

Chuck Reeves
  [8 minutes ago](https://vonage.slack.com/archives/C0124P90QBE/p1689176508631139?thread_ts=1689173099.181669&cid=C0124P90QBE)
If they are using the specific product package then they know which type of auth that product demands and they will set it up correctly

## Sending MMS Using Node.js

Now that the admin is complete, let's create a Node.js script that will use the [Vonage Server SDK for Node.js](https://github.com/vonage/vonage-node-sdk) to send an MMS to a pre-set number when run.

Start by creating a new folder (and if you haven't already done so, put the `private.key` file generated earlier in it.

> Run the following commands to set up a new Node.js application and to install the `vonage-node` client library:

Reviewer 3: I would just add, "From the terminal..."

```
npm init -y
npm install -g @vonage/cli
npm install @vonage/server-sdk
```

Then create a new file called `send-mms.js`, and open it inside your editor.

Initialise a new Vonage instance:

```
const { Vonage } = require("@vonage/server-sdk")
const { MMSImage } = require("@vonage/messages")

> 

Reviewer: See earlier comment re API key and secret.

const vonage = new Vonage({
  apiKey: VONAGE_API_KEY, // Found in your Vonage Dashboard
  apiSecret: VONAGE_API_SECRET, // Found in your Vonage Dashboard
  applicationId: VONAGE_APPLICATION_ID, // Generated earlier

>   privateKey: VONAGE_APPLICATION_PRIVATE_KEY_PATH // Generated earlier

Reviewer 3: I would add an example of how this looks. A lot of people have trouble writing paths.

Reviewer 2: I've added it to the comment.

})
```

Finally, add the code to send the MMS message using the Messages API:

```javascript
const image = new MMSImage({
  image: {
    url: "https://placekitten.com/200/30",
  },
  to: "TO_NUMBER",
  from: "FROM_NUMBER",
});

> 

Reviewer: MMS messages also support `caption` now. Might be worth adding this.

Reviewer 2: You'd think it's best to add an example with a caption included or a paragraph saying that?

Reviewer: I'd include `caption` in the example.

Reviewer 2: I've added a caption as requested. @superchilled. I've tested with @mbcrump  phone number and it works.

vonage.messages
  .send(image)
  .then((resp) => console.log(resp))
  .catch((err) => console.error(error));
```

The above code sends a single message (a picture of a cat) to any number you add into the first object. If for any reason there is a failure in sending the message, the issues will be logged to the console, otherwise, if everything works as it should then the `messageUUID` will be logged instead.

**N﻿ote**: 'https://placekitten.com/200/300', // This is a placeholder image you can use.

With that, head to your terminal of choice and run:

```
node send-mms.js
```

## Where To Go From Here

> Sometimes, for any number of reasons, an MMS message may not be delivered to a recipient so, as a next step from this project, you could experiment with our [Dispatch API](https://developer.vonage.com/en/dispatch/overview) and build a failover mechanism that checks whether the MMS has been delivered, and if not, your app would instead send a standard SMS message with a link to the image you wanted to include.

Reviewer 3: Don't forget to mention Twitter/Slack. Something like:

Let us know what you're building with MMS. Come join the conversation on our [Vonage Community Slack](https://developer.vonage.com/community/slack) or send us a message on [Twitter](https://twitter.com/VonageDev).

---
title: How to Send MMS Messages with Node.js

> description: The Vonage Messages allows you to send MMS (Multimedia Messaging

Reviewer: ```suggestion
description: The Vonage Messages API allows you to send MMS (Multimedia Messaging
```

Reviewer 2: Good catch, thanks.

>   Service) from US shortcode numbers you have in your dashboard to other numbers

Reviewer: ```suggestion
  Service) from US shortcode, 10DLC, and TFN numbers you have in your dashboard to other numbers
```
